include/mbedtls/ssl.h

```diff
@@ -1586,6 +1610,41 @@ typedef int mbedtls_ssl_export_keys_t( void *p_expkey,
                                 size_t maclen,
                                 size_t keylen,
                                 size_t ivlen );
+
+/**
+ * \brief           Callback type: Export key block, master secret,
+ *                                 handshake randbytes and the tls_prf function
+ *                                 used to derive keys.
+ *
+ * \note            This is required for certain uses of TLS, e.g. EAP-TLS
+ *                  (RFC 5216) and Thread. The key pointers are ephemeral and
+ *                  therefore must not be stored. The master secret and keys
+ *                  should not be used directly except as an input to a key
+ *                  derivation function.
+ *
+ * \param p_expkey  Context for the callback.
+ * \param ms        Pointer to master secret (fixed length: 48 bytes).
+ * \param kb            Pointer to key block, see RFC 5246 section 6.3.
+ *                      (variable length: 2 * maclen + 2 * keylen + 2 * ivlen).
+ * \param maclen        MAC length.
+ * \param keylen        Key length.
+ * \param ivlen         IV length.
+ * \param tls_prf       The TLS PRF function used in the handshake.
+ * \param client_random The client random bytes.
+ * \param server_random The server random bytes.
+ *
+ * \return          0 if successful, or
+ *                  a specific MBEDTLS_ERR_XXX code.
+ */
+typedef int mbedtls_ssl_export_keys_ext_t( void *p_expkey,
+                                           const unsigned char *ms,
+                                           const unsigned char *kb,
+                                           size_t maclen,
+                                           size_t keylen,
+                                           size_t ivlen,
+                                           mbedtls_ssl_tls_prf *tls_prf,
```

Comment:
I missed this during the design review, but it would be best to define a new enum with constants corresponding to each individual `tls_prf` functions and pass that instead of the function pointer directly. It would be much safer and the interface would be less complicated too.

Reply:
But the user doesn't have access to any of the individual `tls_prf` funciton.

Reply:
Yes, in this case we would need to add a `tls_prf` function to the API, perhaps that takes the enum as an extra parameter and dispatches it to the right function.

This would be the main benefit of the approach: the user wouldn't go around the API with a function pointer, using something undocumented deep in the library that can change any time, but could use a public API that has stability guarantees like any API in the corresponding release.

Reply:
If we have an external `tls_prf` function, let's say `mbedtls_ssl_tls_prf()` then it would either need to receive the `ssl_context` or the enum as parameter, in order to know what actual `tls_prf` to use. I am in favor of using the context, as using the parameter includes additional logic of either:
1. During the configuration, setting the enum as well, in this case the enum should be part of the context.
or
2. During key derivation, setting the enum according to the `tls_prf` function used \ TLS version used. In this case we add some additional precompilation checks, same way they are done when determining the PRf function to use. This makes the code less readable, and might affect a bit performance ( although  I doubt that by much )
If we put the context as aptrameter to the API, then the enum is redundant altogether.

Reply:
Unfortunately, the `ssl->handshake` is set to NULL on handshake wrapup so an enum it is

Reply:
I think the 2. option would be simpler and would come with less API change for now. I agree that passing the context would make the enum redundant, but I hoped we can do it without passing the context.

Reply:
Yes. I already did it. As mentioned in https://github.com/ARMmbed/mbedtls/pull/2538#discussion_r283129790, using an enum is the only option at the moment, wihtout changing implementation without security risks